specs/agents/breaking-changes.md

```diff
@@ -0,0 +1,40 @@
+# What is a breaking change in a version of an APM agent?
+A change is defined as breaking if it causes an application using an APM agent to break or if the APM product is no longer usable in a way that it previously was.
+
+Taken strictly, this definition could lead to treating every change in runtime behavior as a breaking change. At the same time, we need to be able to keep driving improvements to existing features of our APM product. This document gives some guidelines to help distinguish changes of implementation details from breaking changes.
+
+## Types of breaking changes
+### Instrumentation versions
+Each agent instruments a number of libraries that are used in their language ecosystem. These libraries themselves may introduce new versions, breaking changes, and deprecate older versions. The APM agents therefore will occasionally introduce changes in their instrumentation of external libraries. The changes that we consider breaking are ones that remove support for older versions. Agents can also continue supporting the instrumentation of a particular older library version but drop its testing of it because of some conflicts in installing test suite dependencies, for example. This change would not be considered breaking as long as it’s properly documented.
+
+### Language support
+Similar to library version instrumentation, APM agents will typically support multiple versions of its language. Sometimes it is necessary to drop support for older versions of the languages as they themselves are EOL’ed. It is considered a breaking change when an APM agent drops support for a particular language version.
+
+### Configuration changes
+All agents support a set of configuration options and default values. Changes to the configuration offering can be categorized into two types:
+
+__Change in default configuration value__: Each APM agent configuration option has a default value. Sometimes we change what that default configuration value is. We should consider the _effect_ of changing the value when we evaluate whether the change is breaking. For example, the default configuration value could enrich the data and provide an enhanced experience to the user. In this case, we wouldn’t consider the change to be breaking. On the other hand, if a default value is changed, and as a consequence, removes some information the user was previously able to see, we would consider that a breaking change.
```

Comment:
> On the other hand, if a default value is changed, and as a consequence, removes some information the user was previously able to see, we would consider that a breaking change.

By this definition the change in the Node.js APM agent's default value of `captureSpanStackTraces` -- as documented at https://github.com/elastic/apm-agent-nodejs/blob/main/CHANGELOG.asciidoc#3300-20220310 -- would be considered a breaking change. In this case, the argument was that the perf impact of the `captureSpanStackTraces:true` default was so bad that it justified the change without either a deprecation period or a major version bump.

I'm not sure if I'd advocating calling `captureSpanStackTraces:true` a *bug* (because of the perf impact), or that a spec on breaking changes allow for "within reason" arguments for special cases.  I know the latter somewhat undermines the usefulness of having a breaking changes policy.

Or does the "In general, we don’t consider changes in the data we collect to be breaking." below provide an out for this case?

Reply:
This is a good question. I would lean towards considering `captureSpanStackTraces:true` change a bug fix.
In light of the definition above
_A change is defined as breaking if it causes an application using an APM agent to break or if the APM product is no longer usable in a way that it previously was_
this default option change actually improved the agent functionality in 100% (?) of cases.
I would also hesitate to document that there are special exceptions to breaking changes and would rather focus on defining whether something is a bug. What do you think?